presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java

```diff
@@ -23,79 +24,190 @@
 import static com.facebook.presto.verifier.checksum.ArrayColumnValidator.generateArrayChecksum;
+import static com.facebook.presto.verifier.checksum.ArrayColumnValidator.getAsDoubleArrayColumn;
 import static com.facebook.presto.verifier.framework.VerifierUtil.delimitedIdentifier;
 import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
 
 public class MapColumnValidator
         implements ColumnValidator
 {
+    private final FloatingPointColumnValidator floatingPointValidator;
+    private final boolean validateStringAsDouble;
+
+    @Inject
+    public MapColumnValidator(VerifierConfig config, FloatingPointColumnValidator floatingPointValidator)
+    {
+        this.floatingPointValidator = requireNonNull(floatingPointValidator, "floatingPointValidator is null");
+        this.validateStringAsDouble = config.isValidateStringAsDouble();
+    }
+
     @Override
     public List<SingleColumn> generateChecksumColumns(Column column)
     {
         checkArgument(column.getType() instanceof MapType, "Expect MapType, found %s", column.getType().getDisplayName());
         Type keyType = ((MapType) column.getType()).getKeyType();
         Type valueType = ((MapType) column.getType()).getValueType();
+        ImmutableList.Builder<SingleColumn> builder = ImmutableList.builder();
 
         Expression checksum = functionCall("checksum", column.getExpression());
         Expression keysChecksum = generateArrayChecksum(functionCall("map_keys", column.getExpression()), new ArrayType(keyType));
+        // checksum(cardinality(map_column))
         Expression mapCardinalityChecksum = functionCall("checksum", functionCall("cardinality", column.getExpression()));
+        // coalesce(sum(cardinality(map_column)), 0)
         Expression mapCardinalitySum = new CoalesceExpression(
                 functionCall("sum", functionCall("cardinality", column.getExpression())),
                 new LongLiteral("0"));
 
+        builder.add(new SingleColumn(checksum, Optional.of(delimitedIdentifier(getChecksumColumnAlias(column)))));
+        builder.add(new SingleColumn(keysChecksum, Optional.of(delimitedIdentifier(getKeysChecksumColumnAlias(column)))));
+
         // We need values checksum in one case only: when key is a floating point type and value is not.
         // In such case, when both column checksum and the key checksum do not match, we cannot tell if the values match or not.
         // Meaning we cannot resolve column mismatch, because value type is not a floating type and a mismatch in the values could indicate a real correctness issue.
         // In order to resolve column mismatch in such a situation, generate an extra checksum for the values.
-        if (isFloatingPointType(keyType) && !isFloatingPointType(valueType)) {
+        if ((isFloatingPointType(keyType) || shouldValidateStringAsDouble(keyType)) && !isFloatingPointType(valueType)) {
```

Comment:
should we also skip this if the value type is a string that's really a double?

Reply:
Hi Rebecca, as in another comment, after this MapColumnValidator component, there's another related component, called FailureResolver. MapColumnValidator does checksum exact match today instead of floating point validation. When MapColumnValidator returns mismatch, if the FailureResolver sees the map comprised of floating point, the FailureResolver auto resolves the mismatch and considers it correct.

The reason that we skip here today for floating point is because of this FailureResolver approach. And I think we should get rid of FailureResolver and make MapColumnValidator do floating point validation, as ArrayColumnValidator does wtih https://github.com/prestodb/presto/pull/22143, for floating point.

In this PR for map of varchar, I directly do floating point validation for varchar. So we need to check on both key column and value column, during which we do floating point validation if either of them is a floating point column or a varchar column that's really a double.